### PR TITLE
[inductor] post_grad batched linear fusion

### PR DESCRIPTION
**Pull request by xuzhao9**
Summary: Fusing independent nn.Linear() functions with aten.bmm and aten.cat.

Test Plan:
Without the BMM fusion:
```
buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 0
```
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072536_6535183793.json.gz&bucket=pyper_traces

100 aten::mm operators

With the BMM fusion:
```
buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1
```

20 aten::bmm operators

https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/torchbench_test_module_20231030_072157_6535183793.json.gz&bucket=pyper_traces

Passes accuracy test:
```
$ buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --accuracy
Running eval method from test_module on cuda in dynamo inductor mode with input batch size 4 and precision tf32.
Accuracy:                            pass
```
Looks like the bmm and input cat has been fused successfully.

Checking the triton codegen:

```
TORCH_LOGS=+dynamo,+aot,+inductor buck2 run @mode/opt //pytorch/benchmark:run -- test_module -d cuda --module test_linear_module --torchdynamo inductor --torchinductor_cudagraph 0 --torchinductor_batch_fusion 1 --dump_triton 1
```

Triton code dump: https://www.internalfb.com/intern/everpaste/?handle=GHp1ABaqYuTjYCUBALiTWmteaI1PbsIXAAAB

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by xuzhao9** [comment]
@yanboliang Can I get a review on this PR?

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -79,6 +79,9 @@
 # enable pattern match with batch fusion (using torch op)
 batch_fusion = True
 
+# enable post-grad pattern match with batch fusion (using torch op)
+post_grad_batch_fusion = False
+
```
Comment (yanboliang) [inline torch/_inductor/config.py:84]: I don't think we have to add a separate flag for this, we already have ```batch_fusion``` at L80.

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 2]
```diff
@@ -66,6 +66,89 @@ class BatchFusion(GroupBatchFusionBase):
     pass
 
 
+class BatchLinearPostGradFusion(GroupBatchFusionBase):
```
Comment (yanboliang) [inline torch/_inductor/fx_passes/group_batch_fusion.py:69]: Probably we have to update the original ```BatchLinearFusion``` to ```PreGradBatchLinearFusion``` and name this as ```PostGradBatchLinearFusion```.

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 3]
```diff
@@ -653,6 +739,9 @@ def group_batch_fusion_post_grad_passes(graph: torch.fx.Graph):
     if config.group_fusion and has_fbgemm:
         fusions += [GroupLinearFusion()]
 
+    if config.post_grad_batch_fusion:
+        fusions += [BatchLinearPostGradFusion()]
+
```
Comment (yanboliang) [inline torch/_inductor/fx_passes/group_batch_fusion.py:744]: I think we need a mechanism to only enable one batch linear fusion, either pre or post grad.

Comment (yanboliang) [reply torch/_inductor/fx_passes/group_batch_fusion.py:744]: Or probably we have a config can be used to specify a list of fusions and the order.

File: torch/_inductor/config.py (hunk 1 above)
Comment (xuzhao9) [reply torch/_inductor/config.py:84]: I thought `batch_fusion` at L80 is pre_grad? I am thinking it would still be useful to make them separate config flags for debugging purpose.

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 4]
```diff
@@ -722,14 +806,10 @@ def group_batch_fusion_passes(graph: torch.fx.Graph, pre_grad=True):
...
-            config.pre_grad_fusion_options, pre_grad=True
-        )
-    elif has_fbgemm:  # Only group fusion (which needs fbgemm) in post grad.
-        fusions = generate_fusion_from_config(
-            config.post_grad_fusion_options, pre_grad=False
-        )
+    fusion_options = (
+        config.pre_grad_fusion_options if pre_grad else config.post_grad_fusion_options
+    )
+    fusions = generate_fusion_from_config(fusion_options, pre_grad=pre_grad)
```
Comment (yanboliang) [inline torch/_inductor/fx_passes/group_batch_fusion.py:812]: We need to check ```has_fbgemm```, as we can't run group fusion if fbgemm is not available. I'd like to add a ```need_fbgemm``` attribute to ```GroupBatchFusionBase``` and enable the fusion pass if the check passed.

Comment (xuzhao9) [reply torch/_inductor/fx_passes/group_batch_fusion.py:812]: Right, I will fix this on top of https://github.com/pytorch/pytorch/pull/114778

Comment (xuzhao9) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: Hi @yanboliang , I have addressed your comment, can you help take another look?

**Review by yanboliang — APPROVED** [review]
